aten/src/ATen/core/ivalue.h

```diff
@@ -1228,8 +1228,23 @@ getCustomClassTypeMap();
 template <typename T>
 c10::ClassTypePtr getCustomClassTypeImpl() {
   auto& tmap = c10::getCustomClassTypeMap();
-  auto res = tmap.find(std::type_index(typeid(T)));
-  TORCH_CHECK(res != tmap.end(), "Can't find class id in custom class type map", "");
+  auto tindex = std::type_index(typeid(T));
+  auto res = tmap.find(tindex);
+  if (C10_UNLIKELY(res == tmap.end())) {
+    // type_index is not guaranteed to be unique across shared libraries on some platforms
+    // For example see https://github.com/llvm-mirror/libcxx/blob/78d6a7767ed57b50122a161b91f59f19c9bd0d19/include/typeinfo#L133
+    // Also, this is not the case if RTLD_LOCAL option is used, see
+    // https://github.com/pybind/pybind11/blob/f791dc8648e1f6ec33f402d679b6b116a76d4e1b/include/pybind11/detail/internals.h#L101-L106
+    // Take a slow path of iterating over all registered types and compare their names
+    auto class_name = std::string(tindex.name());
+    for(const auto &it: tmap) {
+      if (class_name == it.first.name()) {
+          tmap[tindex] = it.second;
```

Comment:
Are the any race condition risks by modifying the map in-place? It doesn't seem the map is guarded by a mutex but it was probably fine until now since I believe insertions happened during static initialization (and after we only ever read from the map). If we allow the map to be modified as we go this could lead to corruption if two threads do it simultaneously.

Reply:
Good point, will do as follow up PR (as currently it is only triggered on macOS, and will probably require some design decisions)